File: sdk/client/cycles_test.go
package client

import (
	"context"
	"encoding/json"
	"net/http"
	"net/http/httptest"
	"testing"
)

func TestListCycles_WrappedShape(t *testing.T) {
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/api/v1/projects/5/cycles" {
			t.Errorf("unexpected path %s", r.URL.Path)
		}
		if r.URL.Query().Get("status") != "active" {
			t.Errorf("unexpected query %s", r.URL.RawQuery)
		}
		json.NewEncoder(w).Encode(map[string]any{
			"items":  []map[string]any{{"id": 1, "name": "Sprint 1", "status": "active", "progress": 40.5}},
			"total":  1, "limit": 50, "offset": 0,
		})
	}))
	defer srv.Close()

	c := New(srv.URL+"/api/v1", "t")
	res, err := c.ListCycles(context.Background(), 5, "active", 50, 0)
	if err != nil || res.Total != 1 || len(res.Items) != 1 || res.Items[0].Progress != 40.5 {
		t.Fatalf("ListCycles: %v %+v", err, res)
	}
}

func TestGetCycleBurndown(t *testing.T) {
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/api/v1/cycles/3/burndown" {
			t.Errorf("unexpected path %s", r.URL.Path)
		}
		json.NewEncoder(w).Encode(map[string]any{
			"cycle_id": 3, "cycle_name": "S1", "total_issues": 10, "is_on_track": true,
			"daily_points": []map[string]any{{"day_index": 0, "actual_remaining": 9.0}},
		})
	}))
	defer srv.Close()

	c := New(srv.URL+"/api/v1", "t")
	b, err := c.GetCycleBurndown(context.Background(), 3)
	if err != nil || b.IsOnTrack != true || len(b.DailyPoints) != 1 {
		t.Fatalf("GetCycleBurndown: %v %+v", err, b)
	}
}

func TestAddIssueToCycle(t *testing.T) {
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/api/v1/cycles/3/issues" {
			t.Errorf("unexpected path %s", r.URL.Path)
		}
		if r.URL.Query().Get("issue_id") != "11" {
			t.Errorf("unexpected query %s", r.URL.RawQuery)
		}
		json.NewEncoder(w).Encode(map[string]any{"cycle_id": 3, "issue_id": 11, "action": "added"})
	}))
	defer srv.Close()

	c := New(srv.URL+"/api/v1", "t")
	if err := c.AddIssueToCycle(context.Background(), 3, 11); err != nil {
		t.Fatalf("AddIssueToCycle: %v", err)
	}
}

func TestGetCycle(t *testing.T) {
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/api/v1/cycles/3" {
			t.Errorf("unexpected path %s", r.URL.Path)
		}
		json.NewEncoder(w).Encode(map[string]any{
			"id": 3, "name": "Sprint 1", "status": "active", "progress": 40.5,
			"total_issues": 10, "completed_issues": 4, "start_date": "2026-09-01",
		})
	}))
	defer srv.Close()

	c := New(srv.URL+"/api/v1", "t")
	cy, err := c.GetCycle(context.Background(), 3)
	if err != nil || cy.Name != "Sprint 1" || cy.TotalIssues != 10 {
		t.Fatalf("GetCycle: %v %+v", err, cy)
	}
}

func TestGetCycleProgress(t *testing.T) {
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/api/v1/cycles/3/progress" {
			t.Errorf("unexpected path %s", r.URL.Path)
		}
		json.NewEncoder(w).Encode(map[string]any{
			"cycle_id": 3, "cycle_name": "Sprint 1", "total_issues": 10, "completed_issues": 4, "progress": 40.5,
			"state_breakdown": []map[string]any{{"state": "Done", "group": "completed", "count": 4}},
		})
	}))
	defer srv.Close()

	c := New(srv.URL+"/api/v1", "t")
	p, err := c.GetCycleProgress(context.Background(), 3)
	if err != nil || p.Progress != 40.5 || len(p.StateBreakdown) != 1 || p.StateBreakdown[0].State != "Done" {
		t.Fatalf("GetCycleProgress: %v %+v", err, p)
	}
}

File: sdk/node/tests/cycles.test.ts
import { describe, it, expect } from "vitest";
import { createServer } from "node:http";
import { ReqMangoClient } from "../src/index.js";

function mockServer(handler: (req: any, res: any) => void): Promise<{ url: string; close: () => void }> {
  return new Promise((resolve) => {
    const srv = createServer(handler);
    srv.listen(0, "127.0.0.1", () => {
      const addr = srv.address() as any;
      resolve({
        url: `http://127.0.0.1:${addr.port}/api/v1`,
        close: () => srv.close(),
      });
    });
  });
}

describe("Cycles", () => {
  it("listCycles decodes wrapped shape", async () => {
    const srv = await mockServer((_req, res) => {
      res.writeHead(200, { "Content-Type": "application/json" });
      res.end(JSON.stringify({
        items: [{ id: 1, name: "Sprint 1", status: "active", progress: 40.5 }],
        total: 1, limit: 50, offset: 0,
      }));
    });
    try {
      const client = new ReqMangoClient({ baseUrl: srv.url, token: "t" });
      const result = await client.listCycles(5, { status: "active" });
      expect(result.total).toBe(1);
      expect(result.items).toHaveLength(1);
      expect(result.items[0].progress).toBe(40.5);
    } finally {
      srv.close();
    }
  });

  it("getCycle decodes single object", async () => {
    const srv = await mockServer((_req, res) => {
      res.writeHead(200, { "Content-Type": "application/json" });
      res.end(JSON.stringify({ id: 3, name: "Sprint 1", status: "active" }));
    });
    try {
      const client = new ReqMangoClient({ baseUrl: srv.url, token: "t" });
      const cycle = await client.getCycle(3);
      expect(cycle.id).toBe(3);
      expect(cycle.name).toBe("Sprint 1");
    } finally {
      srv.close();
    }
  });

  it("getCycleBurndown decodes daily points", async () => {
    const srv = await mockServer((_req, res) => {
      res.writeHead(200, { "Content-Type": "application/json" });
      res.end(JSON.stringify({
        cycle_id: 3, cycle_name: "S1", total_issues: 10, is_on_track: true,
        daily_points: [{ day_index: 0, actual_remaining: 9.0 }],
      }));
    });
    try {
      const client = new ReqMangoClient({ baseUrl: srv.url, token: "t" });
      const b = await client.getCycleBurndown(3);
      expect(b.is_on_track).toBe(true);
      expect(b.daily_points).toHaveLength(1);
    } finally {
      srv.close();
    }
  });

  it("addIssueToCycle sends POST", async () => {
    const srv = await mockServer((req, res) => {
      expect(req.method).toBe("POST");
      expect(req.url).toContain("issue_id=11");
      res.writeHead(200, { "Content-Type": "application/json" });
      res.end(JSON.stringify({ cycle_id: 3, issue_id: 11, action: "added" }));
    });
    try {
      const client = new ReqMangoClient({ baseUrl: srv.url, token: "t" });
      await client.addIssueToCycle(3, 11); // should not throw
    } finally {
      srv.close();
    }
  });
});
